### PR TITLE
Make choosing an executor (native/wasm) an explicit part of service construction

### PR DESCRIPTION
**Pull request by expenses**
Polkadot companion: https://github.com/paritytech/polkadot/pull/3615

### CODE REVIEW

**Comment by expenses** [comment]
Fully closes https://github.com/paritytech/substrate/issues/9475 .

File: bin/node-template/node/src/service.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ native_executor_instance!(
 	frame_benchmarking::benchmarking::HostFunctions,
 );
 
-type FullClient = sc_service::TFullClient<Block, RuntimeApi, Executor>;
+type FullClient = sc_service::TFullClient<Block, RuntimeApi, NativeExecutor<Executor>>;
```
Comment (expenses) [inline bin/node-template/node/src/service.rs:24]: We should probably rename `Executor` here to something like `ExecutorDispatch`.

File: bin/node/cli/src/service.rs [hunk 2]
```diff
@@ -26,18 +26,19 @@ use node_primitives::Block;
...
 use sc_consensus_babe::{self, SlotProportion};
+use sc_executor::NativeExecutor;
 use sc_network::{Event, NetworkService};
 use sc_service::{config::Configuration, error::Error as ServiceError, RpcHandlers, TaskManager};
 use sc_telemetry::{Telemetry, TelemetryWorker};
 use sp_runtime::traits::Block as BlockT;
 use std::sync::Arc;
 
-type FullClient = sc_service::TFullClient<Block, RuntimeApi, Executor>;
+type FullClient = sc_service::TFullClient<Block, RuntimeApi, NativeExecutor<Executor>>;
```
Comment (gilescope) [inline bin/node/cli/src/service.rs:36]: Are we tying ourselves more to a Native Executor here? (Presumably there will be clients that want to just use a wasm executor?)

Comment (expenses) [reply bin/node/cli/src/service.rs:36]: We're trying the node into the native executor, but other clients can just set the type definition differently

File: test-utils/test-runner/src/client.rs [hunk 3]
```diff
@@ -105,8 +106,14 @@ where
 			default_config(task_executor, chain_spec),
 	};
 
+	let executor = NativeExecutor::<T::Executor>::new(
```
Comment (chevdor) [inline test-utils/test-runner/src/client.rs:109]: I find the naming a little confusing there, passing WASM data to create a **Native**Executor.

**Review by chevdor — CHANGES_REQUESTED** [review]
LGTM besides the naming issue

**Comment by expenses** [comment]
The majority of this PR now deals with a renaming of:

* `NativeExecutor` -> `NativeElseWasmExecutor`
* `Executor` -> `ExecutorDispatch`

File: bin/node/executor/src/lib.rs [hunk 4]
```diff
@@ -22,9 +22,9 @@ pub use sc_executor::NativeExecutor;
 
 // Declare an instance of the native executor named `Executor`. Include the wasm binary as the
```
Comment (gilescope) [inline bin/node/executor/src/lib.rs:23]:
```suggestion
// Declare an instance of the native executor named `ExecutorDispatch`. Include the wasm binary as the
```

**Review by gilescope — APPROVED** [review]

**Review by chevdor — APPROVED** [review]

**Comment by expenses** [comment]
bot merge

**Comment by expenses** [comment]
bot merge

**Comment by expenses** [comment]
bot merge

**Comment by expenses** [comment]
bot merge

**Comment by expenses** [comment]
Fuck it

**Pull request merged**